### PR TITLE
Add hyperparameter search feature

### PR DESCRIPTION
**Pull request by Mouhanedg56**
#83 implementation.

Add hyperparameter search method to `SetfitTrainer`. Only `optuna` is supported for now because other backends need major changes in `SetfitTrainer` attributes.
There are two hyperparameter types:

- Training arguments such as seed, epoch, batch_size, learning rate, ... to be listed using `hp_space`.
- Model head parameters: in our case the head is a logistic regression, to be listed using `model_init`.

This is how to set up and perform the hyperparameter search for both parameters:
```python
from datasets import Dataset

from setfit.trainer import SetFitTrainer
from setfit.modeling import SetFitModel

def hp_space(trial):  # Training arguments
    return {
        "learning_rate": trial.suggest_float("learning_rate", 1e-6, 1e-4, log=True),
        "batch_size": trial.suggest_categorical("batch_size", [4, 8, 16, 32, 64]),
    }

def model_init(trial):  # Model head parameters
    if trial is not None:
        max_iter = trial.suggest_int("max_iter", 50, 300)
        solver = trial.suggest_categorical("solver", ["newton-cg", "lbfgs", "liblinear"])
    else:
        max_iter = 100
        solver = "liblinear"
    params = {
        "head_params": {
            "max_iter": max_iter,
            "solver": solver,
        }
    }
    return SetFitModel.from_pretrained("sentence-transformers/paraphrase-albert-small-v2", **params)

dataset = Dataset.from_dict(
            {"text_new": ["a", "b", "c"], "label_new": [0, 1, 2], "extra_column": ["d", "e", "f"]}
        )

trainer = SetFitTrainer(
    train_dataset=dataset,
    eval_dataset=dataset,
    model_init=model_init,
    column_mapping={"text_new": "text", "label_new": "label"},
)
trainer.hyperparameter_search(direction="minimize", hp_space=hp_space, n_trials=4)
```

### CODE REVIEW

File: src/setfit/trainer.py [hunk 1]
```diff
@@ -224,17 +276,94 @@ def train(self):
...
+        hp_space: Optional[Callable[["optuna.Trial"], Dict[str, float]]] = None,
+        compute_objective: Optional[Callable[[Dict[str, float]], float]] = None,
+        n_trials: int = 10,
+        direction: str = "minimize",
+        backend: Optional[Union["str", HPSearchBackend]] = None,
+        hp_name: Optional[Callable[["optuna.Trial"], str]] = None,
+        **kwargs,
+    ) -> BestRun:
+        """
+        Launch an hyperparameter search using `optuna`. The optimized quantity is determined
```
Comment (lewtun) [inline src/setfit/trainer.py:302]: turbo nit:
```suggestion
        Launch a hyperparameter search using `optuna`. The optimized quantity is determined
```

File: src/setfit/trainer.py [hunk 2]
```diff
@@ -224,17 +276,94 @@ def train(self):
...
+        by `compute_objective`, which defaults to a function returning the evaluation loss when no metric is provided,
+        the sum of all metrics otherwise.
+
+        <Tip warning={true}>
+
+        To use this method, you need to have provided a `model_init` when initializing your [`Trainer`]: we need to
```
Comment (lewtun) [inline src/setfit/trainer.py:308]: nit:
```suggestion
        To use this method, you need to have provided a `model_init` when initializing your [`SetFitTrainer`]: we need to
```

File: src/setfit/utils.py [hunk 3]
```diff
@@ -49,6 +49,18 @@
 }
 
 
+def default_hp_space_optuna(trial) -> Dict[str, float]:
+    from transformers.integrations import is_optuna_available
+
+    assert is_optuna_available(), "This function needs Optuna installed: `pip install optuna`"
+    return {
+        "learning_rate": trial.suggest_float("learning_rate", 1e-6, 1e-4, log=True),
+        "num_epochs": trial.suggest_int("num_epochs", 1, 5),
```
Comment (lewtun) [inline src/setfit/utils.py:58]: What do you think about adding `num_iterations` (i.e. the number of contrastive learning triples to generate) to the default space, since this hyperparameter is quite important for SetFit?

E.g. we could include:

```
num_iterations: trial.suggest_categorical("num_iterations", [5,10,20])

File: src/setfit/trainer.py [hunk 4]
```diff
@@ -131,33 +146,70 @@ def _apply_column_mapping(self, dataset: "Dataset", column_mapping: Dict[str, st
         )
         return dataset
 
-    def call_model_init(self):
+    def _hp_search_setup(self, trial: Union["optuna.Trial", Dict[str, Any]]):
+        """HP search setup code"""
```
Comment (lewtun) [inline src/setfit/trainer.py:150]: Let's add a note to the source inspiration :)

```suggestion
        """HP search setup code"""
        
         # Heavily inspired by transformers.Trainer._hp_search_setup
         # https://github.com/huggingface/transformers/blob/cbb8a37929c3860210f95c9ec99b8b84b8cf57a1/src/transformers/trainer.py#L1163
```

File: src/setfit/integrations.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    return importlib.util.find_spec("optuna") is not None
+
+
+def default_hp_search_backend():
+    if is_optuna_available():
+        return "optuna"
+
+
+def run_hp_search_optuna(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    import optuna
```
Comment (lewtun) [inline src/setfit/integrations.py:16]: Let's add a note to the source inspiration :)

```suggestion
    import optuna
    
    # Heavily inspired by transformers.integrations.run_hp_search_optuna
    # https://github.com/huggingface/transformers/blob/cbb8a37929c3860210f95c9ec99b8b84b8cf57a1/src/transformers/integrations.py#L160
```

**Review by lewtun — COMMENTED** [review]
Wow, this is an incredible integration @Mouhanedg56 🔥 !!

I've left a few nits and suggestions for the default hyperparameter space. Would you also like to add your example in the PR to the README? It could be a section under the training a model one, with something like "Running hyperparameter search".

This way, people can learn about the new feature since we don't (yet) have proper docs :)

**Comment by lewtun** [comment]
Hey @Mouhanedg56 we just merged #69 which introduced some merge conflicts in your PR. Would you mind resolving them so we can get this nice feature included too?

**Comment by lewtun** [comment]
Thanks for resolving the conflicts! I think you just need to run `make style` on your code and this should be good to go!

**Comment by Mouhanedg56** [comment]
OOPS, I forget to `make style`. Now it should be fixed.
